include/flang/lower/ASTBuilder.h

```diff
@@ -0,0 +1,332 @@
+    static_assert(AST::isConstruct<A>(), "must be a construct");
+  }
+
+  /// is `A` executable (an action statement or compiler generated)?
+  template <typename A>
+  constexpr static bool isAction(const A &a) {
+    return !AST::isConstruct<A>() && !isOther(a);
+  }
+
+  /// is `A` a compiler-generated evaluation?
+  template <typename A>
+  constexpr static bool isGenerated(const A &) {
+    return std::is_same_v<A, CGJump>;
+  }
+
+  /// is `A` not an executable statement?
+  template <typename A>
+  constexpr static bool isOther(const A &) {
+    return std::is_same_v<A, parser::FormatStmt> ||
+           std::is_same_v<A, parser::EntryStmt> ||
+           std::is_same_v<A, parser::DataStmt> ||
+           std::is_same_v<A, parser::NamelistStmt>;
+  }
+
+  constexpr bool isActionStmt() const {
+    return std::visit(common::visitors{
+                          [](auto *p) { return isAction(*p); },
+                          [](auto &r) { return isGenerated(r); },
+                      },
+                      u);
+  }
+
+  constexpr bool isStmt() const {
+    return std::visit(common::visitors{
+                          [](auto *p) { return isAction(*p) || isOther(*p); },
+                          [](auto &r) { return isGenerated(r); },
+                      },
+                      u);
+  }
+  constexpr bool isConstruct() const { return !isStmt(); }
+
+  /// Set the type of originating control flow type for this evaluation.
+  void setCFG(CFGAnnotation a, Evaluation *cstr) {
+    cfg = a;
+    setBranches(cstr);
+  }
+
+  /// Is this evaluation a control-flow origin? (The AST must be annotated)
+  bool isControlOrigin() const { return cfg != CFGAnnotation::None; }
+
+  /// Is this evaluation a control-flow target? (The AST must be annotated)
+  bool isControlTarget() const { return isTarget; }
+
+  /// Set the containsBranches flag iff this evaluation (a construct) contains
+  /// control flow
+  void setBranches() { containsBranches = true; }
+
+  constexpr std::list<Evaluation> *getConstructEvals() {
+    return isStmt() ? nullptr : subs;
+  }
+
+  /// Set that the construct `cstr` (if not a nullptr) has branches.
+  static void setBranches(Evaluation *cstr) {
+    if (cstr) {
+      cstr->setBranches();
+    }
+  }
+
+  EvalVariant u;
+  ParentType parent;
+  parser::CharBlock pos;
+  std::optional<parser::Label> lab;
+  std::list<Evaluation> *subs{nullptr}; // construct sub-statements
+  CFGAnnotation cfg{CFGAnnotation::None};
+  bool isTarget{false};         // this evaluation is a control target
+  bool containsBranches{false}; // construct contains branches
+};
+
+/// A program is a list of program units.
+/// These units can be function like, module like, or block data
+struct ProgramUnit {
+  template <typename A>
+  ProgramUnit(A *ptr, const ParentType &parent) : p{ptr}, parent{parent} {}
+
+  std::variant<const parser::MainProgram *, const parser::FunctionSubprogram *,
```

Comment:
Same as for the other parse-tree pointers, they are inside a struct that can ensure they are not null and hide the pointer aspects. I made the member a constant to strengthen the guarantee it will never be null given the ctors take references.